pkg/shell/cockpit-accounts.js

```diff
@@ -840,88 +897,80 @@ PageAccountSetPassword.prototype = {
         $('#account-set-password-pw1').val("");
         $('#account-set-password-pw2').val("");
-        $('#account-set-password-message-password-mismatch').css("visibility", "hidden");
-        $("#account-set-password-dialog .check-passwords").removeClass("has-error");
-        this.update ();
+        $('#account-set-password-meter').removeClass("weak okay good excellent");
+        $("#account-set-password-dialog").dialog("failure", null);
     },
 
     leave: function() {
     },
 
-    update: function(behavior) {
-        function check_params ()
-        {
-            return ($('#account-set-password-pw1').val() !== "" &&
-                    $('#account-set-password-pw2').val() == $('#account-set-password-pw1').val());
-        }
-
-        function highlight_error() {
-            $("#account-set-password-dialog .check-passwords").addClass("has-error");
-            $('#account-set-password-message-password-mismatch').css("visibility", "visible");
-        }
-
-        function hide_error() {
-            $("#account-set-password-dialog .check-passwords").removeClass("has-error");
-            $('#account-set-password-message-password-mismatch').css("visibility", "hidden");
+    validate: function() {
+        var ex;
+        var pw = $('#account-set-password-pw1').val();
+        if ($('#account-set-password-pw2').val() != pw) {
+            ex = new Error(_("The passwords do not match"));
+            ex.target = "#account-set-password-pw2";
         }
 
-        function check_password_match() {
-            if ($('#account-set-password-pw2').val() !== "" &&
-                $('#account-set-password-pw1').val() !== $('#account-set-password-pw2').val())
-                highlight_error();
-            else
-                hide_error();
-        }
-
-        window.clearTimeout(this.error_timeout);
-        this.error_timeout = null;
-
-        if (behavior == "changeFocus") {
-            if ($('#account-set-password-pw2').val() !== "" &&
-                $('#account-set-password-pw1').val() !== $('#account-set-password-pw2').val())
-                highlight_error();
-            else
-                hide_error();
-        } else if (behavior == "input") {
-            if ($('#account-set-password-pw2').val() !== "" &&
-                $('#account-set-password-pw1').val().indexOf($('#account-set-password-pw2').val()) !== 0)
-                highlight_error();
-            else
-                hide_error();
+        var dfd = $.Deferred();
+        if (ex)
+            dfd.reject(ex);
+        else
+            dfd.resolve();
 
-            this.error_timeout = window.setTimeout(check_password_match, 2000);
-            this.setTimeout = null;
-        }
+        var promise = password_quality(pw)
+            .fail(function(ex) {
+                ex.target = "#account-set-password-pw2";
+            })
+            .always(function(arg) {
+                var strength = this.state() == "resolved" ? arg : "weak";
```

Comment:
it is logical, but I believe the statement would gain clarity from using at least one set of parentheses for the ?-condition